File: import_missing_sections.py
import os
import re
from supabase import create_client, Client

# Initialize Supabase client
url = os.getenv('SUPABASE_URL')
key = os.getenv('SUPABASE_SERVICE_ROLE_KEY')
client = Client(url, key)

# Read SQL file
sql_file = r"C:\Users\fujitsu\Desktop\fly-book\pvp_sections_clean.sql"
with open(sql_file, 'r', encoding='utf-8') as f:
    content = f.read()

# Get existing order_nums from database
existing_data = client.table('guide_sections').select('order_num').eq('document_id', 1).execute()
existing_nums = {item['order_num'] for item in existing_data.data}
print(f"Existing order_nums: {len(existing_nums)} entries")

# Parse all INSERT statements from SQL
inserts = []
lines = content.split('\n')
current_stmt = ''

for line in lines:
    stripped = line.strip()
    if stripped.startswith('INSERT INTO guide_sections'):
        if current_stmt:
            inserts.append(current_stmt)
        current_stmt = stripped
    elif current_stmt:
        current_stmt += '\n' + line
        if stripped.endswith(';'):
            current_stmt = current_stmt.rstrip().rstrip(';')
            inserts.append(current_stmt)
            current_stmt = ''

print(f"Total INSERT statements in SQL: {len(inserts)}")

# Parse each INSERT to extract order_num
missing_inserts = []
for insert in inserts:
    # Extract order_num from INSERT statement
    match = re.search(r'order_num\)\s*VALUES\s*\(\d+,\s*(\d+|NULL),\s*[\'"]([^\'"]*)[\'"],\s*(\d+)\)', insert)
    if not match:
        match = re.search(r'order_num\)\s*VALUES\s*\((\d+),\s*(\d+|NULL),\s*[\'"]([^\'"]*)[\'"],\s*\$\$[^$]*\$\$,\s*(\d+)\)', insert)

    if match:
        doc_id = int(match.group(1))
        parent_id = match.group(2)
        title = match.group(3)
        order_num = int(match.group(4))

        if doc_id == 1 and order_num not in existing_nums:
            missing_inserts.append((order_num, insert))

print(f"Missing sections: {len(missing_inserts)}")
print(f"Missing order_nums: sorted({[x[0] for x in missing_inserts[:20]]}...)")

# Execute missing inserts
batch_size = 20
imported = 0
failed = 0

for i in range(0, len(missing_inserts), batch_size):
    batch = missing_inserts[i:i+batch_size]
    print(f'Processing batch {i//batch_size + 1}/{(len(missing_inserts)-1)//batch_size + 1}...')

    for order_num, insert in batch:
        try:
            # Execute the raw SQL via Supabase
            result = client.rpc('exec_sql', {'sql_query': insert})
            imported += 1
            print(f"  Inserted order_num {order_num}")
        except Exception as e:
            # Try parsing and inserting via API instead
            try:
                match = re.search(
                    r'INSERT INTO guide_sections \(document_id, parent_id, title, content, order_num\) VALUES '
                    r'\((\d+), ([\d]+|NULL), \'([^\']*)\'(?:, \$\$(.*?)\$\$|, NULL), (\d+)\)',
                    insert, re.DOTALL
                )
                if match:
                    doc_id = int(match.group(1))
                    parent_id = None if match.group(2) == 'NULL' else int(match.group(2))
                    title = match.group(3).replace("''", "'")  # Handle escaped quotes
                    content = match.group(4) if match.group(4) != 'NULL' else None
                    order_num = int(match.group(5))

                    # Insert via API
                    client.table('guide_sections').insert({
                        'document_id': doc_id,
                        'parent_id': parent_id,
                        'title': title,
                        'content': content,
                        'order_num': order_num
                    }).execute()
                    imported += 1
                    print(f"  Inserted order_num {order_num} via API")
            except Exception as e2:
                failed += 1
                print(f'  Error inserting order_num {order_num}: {e2}')

print(f'\nDone! Imported: {imported}, Failed: {failed}')

File: import_missing_sections.js
const fs = require('fs');
const path = require('path');
const { createClient } = require('@supabase/supabase-js');

// Load environment variables from .env file
const envPath = path.join(__dirname, '.env');
if (fs.existsSync(envPath)) {
    const envContent = fs.readFileSync(envPath, 'utf-8');
    envContent.split('\n').forEach(line => {
        const [key, ...valueParts] = line.split('=');
        if (key && valueParts.length > 0) {
            process.env[key.trim()] = valueParts.join('=').trim();
        }
    });
}

// Initialize Supabase client
const supabaseUrl = process.env.SUPABASE_URL;
const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY;

if (!supabaseUrl || !supabaseKey) {
    console.error('Missing SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY in environment');
    process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseKey);

// Read the missing sections SQL file
const sqlFile = path.join(__dirname, 'pvp_sections_missing.sql');
const content = fs.readFileSync(sqlFile, 'utf-8');

// Parse INSERT statements
const inserts = [];
const lines = content.split('\n');
let currentStmt = '';

for (const line of lines) {
    const stripped = line.trim();
    if (stripped.startsWith('INSERT INTO guide_sections')) {
        if (currentStmt) {
            inserts.push(currentStmt);
        }
        currentStmt = stripped;
    } else if (currentStmt) {
        currentStmt += '\n' + line;
        if (stripped.endsWith(';')) {
            currentStmt = currentStmt.trim().replace(/;$/, '');
            inserts.push(currentStmt);
            currentStmt = '';
        }
    }
}

console.log(`Total INSERT statements to process: ${inserts.length}`);

// Helper function to parse INSERT statement
function parseInsert(insertStmt) {
    // Match: VALUES (document_id, parent_id, 'title', $$content$$ or NULL, order_num)
    const match = insertStmt.match(
        /VALUES \((\d+), ([\d]+|NULL), '([^']*)'(?:, \$\$(.*?)\$\$|, NULL), (\d+)\)/s
    );
    if (match) {
        return {
            document_id: parseInt(match[1]),
            parent_id: match[2] === 'NULL' ? null : parseInt(match[2]),
            title: match[3].replace(/''/g, "'"),
            content: match[4] === 'NULL' ? null : match[4],
            order_num: parseInt(match[5])
        };
    }
    return null;
}

// Process in batches
async function importBatch(batch) {
    const results = [];
    for (const insert of batch) {
        const data = parseInsert(insert);
        if (data) {
            try {
                const { error } = await supabase
                    .from('guide_sections')
                    .insert(data);
                if (error) {
                    console.error(`  Error inserting order_num ${data.order_num}:`, error.message);
                    results.push({ order_num: data.order_num, success: false, error: error.message });
                } else {
                    results.push({ order_num: data.order_num, success: true });
                }
            } catch (e) {
                console.error(`  Exception for order_num ${data.order_num}:`, e.message);
                results.push({ order_num: data.order_num, success: false, error: e.message });
            }
        }
    }
    return results;
}

async function main() {
    const batchSize = 20;
    let imported = 0;
    let failed = 0;

    for (let i = 0; i < inserts.length; i += batchSize) {
        const batch = inserts.slice(i, i + batchSize);
        const batchNum = Math.floor(i / batchSize) + 1;
        const totalBatches = Math.ceil(inserts.length / batchSize);

        console.log(`\nProcessing batch ${batchNum}/${totalBatches}...`);

        const results = await importBatch(batch);

        for (const result of results) {
            if (result.success) {
                imported++;
                console.log(`  ✓ Inserted order_num ${result.order_num}`);
            } else {
                failed++;
            }
        }

        // Small delay to avoid rate limiting
        await new Promise(resolve => setTimeout(resolve, 100));
    }

    console.log(`\n=== Summary ===`);
    console.log(`Total to import: ${inserts.length}`);
    console.log(`Successfully imported: ${imported}`);
    console.log(`Failed: ${failed}`);
}

main().catch(console.error);
